### DISCUSSION TITLE
Random utilities keep getting purged from my production bundle

### DISCUSSION THREAD

**Discussion by harry-anderson** [tailwindlabs/tailwindcss#4931] [Help] [answered]
Hi there!

I'm trying to get to the bottom of an issue that has been plaguing my project for a while. I'm using Tailwind CSS everywhere and making heavy use of modifiers like `hover:` and `sm:`.  For the most part, it works well but every so often random utilities get purged and don't appear in my production CSS bundle. Why could this be? See details below :) 

I'm using these package versions:
- `next: latest`
- `autoprefixer: ^10.2.1`
- `postcss: 8.2.4`
- `tailwindcss: 2.0.2`

`tailwind.config.js`
```js
module.exports = {
  purge: ["./components/**/*.tsx", "./pages/**/*.tsx"],
  darkMode: false,
  theme: {
    extend: {},
  },
  variants: {
    extend: {},
  },
  plugins: [require("@tailwindcss/forms")],
};

```
`postcss.config.js`
```js
module.exports = {
  plugins: ['tailwindcss', 'autoprefixer'],
}
```

A simple example of where I'm encountering this issue:

Here is a button with the class `bg-red-700` (among others)
```jsx
<button
    type="button"
    className="inline-flex text-center items-center px-2.5 py-1.5 border border-gray-300 shadow-sm text-xs font-medium rounded-lg text-white bg-red-700 hover:bg-red-800 focus:outline-none focus:ring-2 focus:ring-offset-2 focus:ring-indigo-500"
    onClick={handleDeleteCollaborator}
>
    <span className="flex-grow">Delete</span>
</button>

```

In Development it works just fine, I can see the generated class in my chrome dev tools. (note the red delete button):

![image](https://user-images.githubusercontent.com/14777088/125033450-aa4f2680-e0d2-11eb-9f82-ea50276275cc.png)

Once deployed in production, that `bg-red-700` class has seemly been purged:

![image](https://user-images.githubusercontent.com/14777088/125033686-fdc17480-e0d2-11eb-84ef-07565146027a.png)

The production CSS file does contain some related classes but not `bg-red-700`
```css 

...
.bg-red-100 {
    --tw-bg-opacity: 1;
    background-color: rgba(254,226,226,var(--tw-bg-opacity))
}

.bg-red-500 {
    --tw-bg-opacity: 1;
    background-color: rgba(239,68,68,var(--tw-bg-opacity))
}
...
```

Any ideas of what might be causing this? Thanks in advance :)

**Comment by harry-anderson** [comment, marked as answer]
I was able to fix this issue by:
- upgrading to the latest version of `next` and `tailwind` using `yarn upgrade --latest`.
- adding classes to the `safelist` in my tailwind css config.

Still unsure as to why classes are getting purged out. I believe it is related to the static generation logic in my `nextjs` project and not specifically an issue in the tailwind library.

**Reply by adamwathan** [reply]
Can you share a reproduction repository? You shouldn't have to safelist it if it's in your template files.
